### PR TITLE
support path-style URL for S3 (#134)

### LINKED ISSUE #134: Support path-style URL for S3 (or S3-compatible) storage
**What would you like to be added**:

Support path-style URL for S3 (or S3-compatible) storage

**Why is this needed**:

Currently, JuiceFS only support [virtual hosted-style URI](https://docs.aws.amazon.com/AmazonS3/latest/dev/VirtualHosting.html). The difference between virtual hosted-style and path-style is:

- Virtual hosted-style: `https://<bucket>.s3.<region>.amazonaws.com`
- Path-style: `https://s3.<region>.amazonaws.com/<bucket>`

Although AWS would [deprecate](https://aws.amazon.com/blogs/aws/amazon-s3-path-deprecation-plan-the-rest-of-the-story) path-style URI in the future, but some S3-compatible storage still use path-style (e.g. [Ceph RGW](https://docs.ceph.com/en/latest/radosgw/s3/commons)). So we may need support this type URI.

### PR DESCRIPTION
**Pull request by luohy15**
When uri.Path is not empty, parse uri as path-style, otherwise parse uri as virtual hosted-style.

Passed [TestS3](https://github.com/juicedata/juicefs/blob/5a92bdfa6091042cc1fa21076b5069be1ce14af5/pkg/object/object_storage_test.go#L216) with following values:

```
juicefs-image.s3.ap-northeast-1.amazonaws.com
s3.ap-northeast-1.amazonaws.com/juicefs-image
play.minio.io:9000/juicefs
localhost:443/test
```

Closes #134

### CODE REVIEW

File: pkg/object/s3.go [hunk 1]
```diff
@@ -306,39 +323,46 @@ func newS3(endpoint, accessKey, secretKey string) (ObjectStorage, error) {
 		ep         string
 	)
 
-	if len(hostParts) == 1 { // take endpoint as bucketname
+	if len(hostParts) == 1 {
```
Comment (davies) [inline pkg/object/s3.go:326]: It could be a valid hostname in path-style, for example, http://localhost/bucket

File: pkg/object/s3.go [hunk 2]
```diff
@@ -353,6 +377,8 @@ func newS3(endpoint, accessKey, secretKey string) (ObjectStorage, error) {
 	}
 	if ep != "" {
 		awsConfig.Endpoint = aws.String(ep)
+	}
+	if uri.Path != "" {
```
Comment (davies) [inline pkg/object/s3.go:381]: For backwards compatibility,  we want to use S3ForcePathStyle with a virtual hosted-style bucket.

File: pkg/object/s3.go [hunk 3]
```diff
@@ -306,39 +323,46 @@ func newS3(endpoint, accessKey, secretKey string) (ObjectStorage, error) {
...
-				endpoint = endpoint[len("dualstack."):]
-			}
-			if endpoint == "amazonaws.com" {
-				endpoint = awsDefaultRegion + "." + endpoint
-			}
-			region = strings.Split(endpoint, ".")[0]
-			if region == "external-1" {
+	} else {
+		// get region or endpoint
+		if uri.Path != "" {
```
Comment (davies) [inline pkg/object/s3.go:334]: What will happen if the path is `/` ?

Comment (luohy15) [reply pkg/object/s3.go]: We can trim `/` before check to prevent this situation

File: pkg/object/s3.go (hunk 1 above)
Comment (luohy15) [reply pkg/object/s3.go]: Added

File: pkg/object/s3.go [hunk 4]
```diff
@@ -293,67 +293,91 @@ func autoS3Region(bucketName, accessKey, secretKey string) (string, error) {
...
 
 	ssl := strings.ToLower(uri.Scheme) == "https"
 	awsConfig := &aws.Config{
-		Region:     aws.String(region),
-		DisableSSL: aws.Bool(!ssl),
-		HTTPClient: httpClient,
+		Region:           aws.String(region),
+		DisableSSL:       aws.Bool(!ssl),
+		HTTPClient:       httpClient,
+		S3ForcePathStyle: aws.Bool(true),
```
Comment (davies) [inline pkg/object/s3.go:374]: Can we change to previous behavior? At least, no regression.

Comment (luohy15) [reply pkg/object/s3.go]: I changed it back. It seems we need to add some regression tests.

**Review by davies — APPROVED** [review]
LGTM

**Pull request merged**